db_test.go

```diff
@@ -1301,3 +1301,92 @@ func TestInitializeHeadTimestamp(t *testing.T) {
 		testutil.Equals(t, int64(15000), db.head.MaxTime())
 	})
 }
+
+func TestDB_LabelNames(t *testing.T) {
```

Comment:
wouldn't something like this cover the same test cases, but a a bit more simplified?

```
func TestDB_LabelNames(t *testing.T) {
	tests := []struct {
		title        string
		sampleLabels [][2]string
		exp          [][]string
	}{
		{
			title: "Labels order",
			sampleLabels: [][2]string{
				[2]string{"name1", ""},
				[2]string{"name3", ""},
				[2]string{"name2", ""},
				[2]string{"name4", ""},
				[2]string{"name1", ""},
			},
			exp: [][]string{
				{"name"},
				{"name1", "name3"},
				{"name1", "name2", "name3"},
				{"name1", "name2", "name3", "name4"},
				{"name1", "name2", "name3", "name4"},
			},
		},
	}

	for _, tst := range tests {
		t.Run(tst.title, func(t *testing.T) {
			db, close := openTestDB(t, nil)
			defer close()
			defer db.Close()

			blockRange := DefaultOptions.BlockRanges[0]
			app := db.Appender()
			for i, tuple := range tst.sampleLabels {
				label := labels.FromStrings(tuple[0], tuple[1])
				_, err := app.Add(label, int64(i)*blockRange, 0)
				testutil.Ok(t, err)
				err = app.Commit()
				testutil.Ok(t, err)
				labelNames, err := db.LabelNames()
				testutil.Ok(t, err)
				testutil.Equals(t, tst.exp[i], labelNames)

			}

			// Check the labels after persisting to disk.
			_, err := db.compact()
			testutil.Ok(t, err)
			labelNames, err := db.LabelNames()
			testutil.Ok(t, err)
			testutil.Equals(t, tst.exp[len(tst.exp)-1], labelNames)
		})
	}
}

```

Reply:
Here we are not checking the case where samples are there in both head and disk. So I had kept 2 sets: 1 set in head, 1 set on disk. I will see if there is any better way to test it.

Reply:
ok ping me when ready.

Reply:
@krasi-georgiev I made the changes to make it look more clean. I would still like to have 2 sets of data so that testing head+disk phase is confusion free.

Reply:
yes it is good to test these separately, but how do you know that when calling `db.Labels()` these come from `blockIndexr.LabelNames()` or `headIndexr.LabelNames()`. Why not call these directly and than also call `db.LabelNames()` to make sure it returns the merged results?

Reply:
yes that is important , but in the current test how do you know that the labels come from the head or the disk when calling db.LabelNames() ? 
Why don't we call explicitly headIndexr.LabelNames(), blockIndexr.LabelNames() and db.LabelNames() at the end. 
This way maybe you can apply the same logic even to the simplified test version I suggested.

Reply:
I had split the set in such a way that, each set gives new label names. So when called `db.LabelNames()` (for head+disk case), 1 set of unique names should come from head, and other from disk.

Checking directly from head and block is also good idea. I will add it.

Reply:
yes, but again this is only guessing since you call db.LabelNames() in both cases so best to be more explicit and call directly the head and the block readers.

Reply:
I modified the test (with some renaming and sorting at places). Now I am checking head, block, db separately.

Reply:
Thanks!

sampleLabels1, sampleLabels2 and exp1,exp2 is still bugging me, but right now can't think how to simplify.